noir/noir-repo/aztec_macros/src/transforms/events.rs

```diff
@@ -1,178 +1,306 @@
-            if signature == SIGNATURE_PLACEHOLDER =>
-        {
-            let selector_literal_id = *first_arg_id;
-
-            let structure = interner.get_struct(struct_id);
-            let signature = event_signature(&structure.borrow());
-            interner.update_expression(selector_literal_id, |expr| {
-                *expr = HirExpression::Literal(HirLiteral::Str(signature.clone()));
-            });
-
-            // Also update the type! It might have a different length now than the placeholder.
-            interner.push_expr_type(
-                selector_literal_id,
-                Type::String(Box::new(Type::Constant(signature.len() as u32))),
-            );
-            Ok(())
-        }
-        _ => Err((
-            AztecMacroError::EventError {
-                span: struct_type.borrow().location.span,
-                message: "Signature placeholder literal does not match".to_owned(),
-            },
-            struct_type.borrow().location.file,
-        )),
+
+    let mut sorted_ast = parsed_ast.into_sorted();
+    let serialize_impl = sorted_ast.trait_impls.remove(0);
+
+    Ok(serialize_impl)
+}
+
+fn generate_trait_impl_deserialize(
+    event_type: &str,
+    event_len: u32,
+    event_fields: &[(String, String)],
+) -> Result<NoirTraitImpl, AztecMacroError> {
+    let field_names: Vec<String> = event_fields
+        .iter()
+        .enumerate()
+        .map(|(index, field)| format!("{}: fields[{}]", field.0, index))
+        .collect::<Vec<String>>();
+    let field_input = field_names.join(",");
+
+    let trait_impl_source = format!(
+        "
+    impl dep::aztec::protocol_types::traits::Deserialize<{event_len}> for {event_type} {{
+        fn deserialize(fields: [Field; {event_len}]) -> {event_type} {{
+            {event_type} {{ {field_input} }}
+        }}
+    }}
+    "
+    )
+    .to_string();
+
+    let (parsed_ast, errors) = parse_program(&trait_impl_source);
+    if !errors.is_empty() {
+        dbg!(errors);
+        return Err(AztecMacroError::CouldNotImplementEventInterface {
+            secondary_message: Some(format!("Failed to parse Noir macro code (trait impl of Deserialize for {event_type}). This is either a bug in the compiler or the Noir macro code")),
+        });
+    }
+
+    let mut sorted_ast = parsed_ast.into_sorted();
+    let deserialize_impl = sorted_ast.trait_impls.remove(0);
+
+    Ok(deserialize_impl)
+}
+
+fn generate_fn_get_event_type_id(
+    event_type: &str,
+    field_length: u32,
+) -> Result<NoirFunction, AztecMacroError> {
+    let from_signature_input =
+        std::iter::repeat("Field").take(field_length as usize).collect::<Vec<_>>().join(",");
+    let function_source = format!(
+        "
+        fn get_event_type_id() -> dep::aztec::protocol_types::abis::event_selector::EventSelector {{
+           dep::aztec::protocol_types::abis::event_selector::EventSelector::from_signature(\"{event_type}({from_signature_input})\")
```

Comment:
On this, is noir smart enough to figure it is constant, or will we be computing a keccak whenever we call it? @benesjan in #7066 you computed a keccak in the rs part but might not be needed to introduce if it is smart enough and we have a `NoteSelector` type as well. 

@TomAFrench you are the expert on noir, do you have insights on the constant part of it? 👀

Reply:
Good point. As far as I know constants are pre-computed. Once Tom confirms I will simplify that other PR of mine (will allow me to nuke tiny-keccak dependency).

Reply:
Yeah, Noir will do any constant hashes at compile time.